synapse/handlers/federation_event.py

```diff
@@ -1669,14 +1669,13 @@ async def _auth_and_persist_outliers(
 
         # XXX: it might be possible to kick this process off in parallel with fetching
         # the events.
-        while event_map:
-            # build a list of events whose auth events are not in the queue.
-            roots = tuple(
-                ev
-                for ev in event_map.values()
-                if not any(aid in event_map for aid in ev.auth_event_ids())
-            )
 
+        # We need to persist an event's auth events before the event.
+        auth_graph = {
+            ev: [event_map[e_id] for e_id in ev.auth_event_ids() if e_id in event_map]
```

Comment:
I'm struggling to understand what happens when the event is missing; the graph will be incomplete then, but that's OK because the list of nodes (`event_map.values()`) below is complete.

Reply:
We handle disconnected/incomplete graphs "correctly"  https://github.com/matrix-org/synapse/pull/16649/files#diff-65efdeace8a5973a8ce6038e1b74db4c0a0b70519823df8d67846119fac35738R133-R139, i.e. for a graph `{a: [], b:[]}` we'll return `[a,b]`